src/_bentoml_impl/frameworks/unsloth/__init__.py

```diff
@@ -0,0 +1,187 @@
+    "mistral-instruct",
+    "openchat",
+    "phi-3",
+    "saiga",
+    "solar-instruct",
+    "vicuna",
+    "zephyr",
+]
+
+ModelType = t.Literal["llama", "mistral", "gemma", "gemma2", "qwen2"]
+
+
+def build_bento(
+    model: PreTrainedModel,
+    tokenizer: PreTrainedTokenizerFast,
+    /,
+    model_name: str | None = None,
+    *,
+    chat_template: ChatTemplate,
+    quantization_method: t.Literal["bitsandbytes"] | None = None,
+    save_method: t.Literal["merged_16bit", "merged_4bit"] = "merged_16bit",
+    overwrite: bool = False,
+    service_config: Config | None = None,
+    engine_config: dict[str, t.Any]
+    | None = None,  # arguments to pass to AsyncEngineArgs
+) -> bentoml.Model:
+    # this model is local then model_name must specified, otherwise derived from model_id
+    is_local = getattr(model.config, "_commit_hash", None) is None
+    if is_local is True and model_name is None:
+        raise BentoMLException(
+            'Fine-tune from a local checkpoint requires specifying "model_name".'
+        )
+    else:
+        model_name = model_name or replace_tag(model.config._name_or_path)
+
+    model_type = t.cast(ModelType, model.config.model_type)
+
+    if service_config is None:
+        service_config = {}
+    if engine_config is None:
+        engine_config = {}
+
+    service_config.update({**MAPPINGS[model_type]["service_config"]})
+
+    engine_config.update(MAPPINGS[model.config.model_type]["engine_config"])
+    if quantization_method is not None:
+        engine_config.update(
+            {"quantization": quantization_method, "load_format": quantization_method}
+        )
+
+    try:
+        bentomodel = bentoml.models.get(model_name)
+        if overwrite:
+            raise bentoml.exceptions.NotFound("Overwriting new models.")
+    except bentoml.exceptions.NotFound:
+        with bentoml.models.create(model_name) as bentomodel:
+            model.save_pretrained_merged(
+                bentomodel.path, tokenizer, save_method=save_method
+            )
+
+    build_opts = dict(
+        python=dict(
+            packages=[
+                "pyyaml",
+                "vllm==0.5.5",
+                "fastapi==0.111.0",
+                "unsloth[huggingface] @ git+https://github.com/bentoml/unsloth.git@main",
+            ],
+            lock_packages=True,
+        ),
+        envs=[{"name": "HF_TOKEN"}],
+    )
+    merger.merge(build_opts, get_extras().get(model_type, {}))
+
+    logger.info(
+        "Building bentos for %s, model_id=%s", model_type, model.config._name_or_path
+    )
+
+    with tempfile.TemporaryDirectory() as tempdir:
+        tempdir = pathlib.Path(tempdir)
+        shutil.copytree(
+            pathlib.Path(__file__).parent / "template", tempdir, dirs_exist_ok=True
+        )
+        with (tempdir / "service_config.yaml").open("w") as f:
+            f.write(
+                yaml.safe_dump(
+                    dict(
+                        chat_template=chat_template,
```

Comment:
Why should BentoML be managing the chat template?

Reply:
Otherwise, the prompt won't be correct here. Let say the model is fine-tuned with a specific chat template, then the service should also use that chat_template as override.

Reply:
Chat templates can get updates. It shouldn't be BentoML's responsibility to keep the chat templates update-to-date. Where did we get these chat templates?

Reply:
I believe we got it from https://github.com/chujiezheng/chat_templates

Reply:
When a model is saved, isn't `tokenizer_config.json` saved with the model as well, which includes the bos/eos tokens and chat template? Having BentoML managing the chat template, eos/bos tokens, and system prompts doesn't seem right.

Reply:
Not actually, save_pretrained won't apply the chat templates that user fine-tune the model with. Unless users override the default chat template.

The chat template comes with `tokenizer_config.json` might not be the one that users use to fine tune with.

Reply:
They will have to do something like this: https://huggingface.co/docs/transformers/en/chat_templating#what-template-should-i-use